source/postprocess/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,833 @@
+                                                                  weighted_rotation_matrices[mineral][i_grain]);
+                    }
+                }
+              string_stream_content_draw_volume_weighting << std::endl;
+
+              // write data
+              for (unsigned int grain = 0; grain < n_grains; ++grain)
+                {
+                  string_stream_content_draw_volume_weighting << id << " ";
+                  for (unsigned int property_i = 0; property_i < write_draw_volume_weighted_cpo.size(); ++property_i)
+                    {
+                      switch (write_draw_volume_weighted_cpo[property_i].second)
+                        {
+                          case Output::VolumeFraction:
+                            string_stream_content_draw_volume_weighting << volume_fractions_grains[write_draw_volume_weighted_cpo[property_i].first][grain] << " ";
+                            break;
+
+                          case Output::RotationMatrix:
+                            string_stream_content_draw_volume_weighting << weighted_rotation_matrices[write_draw_volume_weighted_cpo[property_i].first][grain] << " ";
+                            break;
+
+                          case Output::EulerAngles:
+                            Assert(compute_raw_euler_angles == true,
+                                   ExcMessage("Internal error: writing out raw Euler angles, without them being computed."));
+                            string_stream_content_draw_volume_weighting << weighted_euler_angles[write_draw_volume_weighted_cpo[property_i].first][grain][0] << " "
+                                                                        <<  weighted_euler_angles[write_draw_volume_weighted_cpo[property_i].first][grain][1] << " "
+                                                                        <<  weighted_euler_angles[write_draw_volume_weighted_cpo[property_i].first][grain][2] << " ";
+                            break;
+
+                          default:
+                            Assert(false, ExcMessage("Internal error: raw CPO postprocess case not found."));
+                            break;
+                        }
+                    }
+                  string_stream_content_draw_volume_weighting << std::endl;
+                }
+            }
+        }
+
+      std::string filename_master = particle_file_prefix_master + "." + Utilities::int_to_string(dealii::Utilities::MPI::this_mpi_process (MPI_COMM_WORLD),4) + ".dat";
+      std::string filename_raw = particle_file_prefix_content_raw + "." + Utilities::int_to_string(dealii::Utilities::MPI::this_mpi_process (MPI_COMM_WORLD),4) + ".dat";
+      std::string filename_draw_volume_weighting = particle_file_prefix_content_draw_volume_weighting + "." + Utilities::int_to_string(dealii::Utilities::MPI::this_mpi_process (MPI_COMM_WORLD),4) + ".dat";
```

Comment:
Dont assume MPI_COMM_WORLD is the correct communicator, we could have been started with a different one (e.g. as subprocess called from a different MPI program). This is currently not possible, but we want to limit the use of MPI_COMM_WORLD as much as possible to main.cc so that it is easy to extend ASPECT in the future. Use `this->get_mpi_communicator()` instead.